File: website/src/lib/widgetSlug.ts
/**
 * Stable widget-slug derivation.
 *
 * Every `<mcwidget>` impression in chat is bound to an artifact identity
 * via a slug. The agent normally emits an explicit `slug=` attribute when
 * re-rendering a saved artifact; for brand-new emissions the slug is
 * derived deterministically from the message location so that:
 *
 *   - Same message + same widget position → same slug → same binding.
 *   - Save once, refresh, click again → no duplicate created (the second
 *     POST goes to the same slug, server returns 409, frontend reconciles
 *     the bookmark icon to "filled").
 *   - Legacy widgets (rendered before this scheme existed) still get a
 *     stable identity from their message_ts so bookmark state survives
 *     refreshes.
 *
 * The hash function is FNV-1a-like — fast, deterministic, no crypto
 * properties needed (we just want unique-enough opaque IDs). Output is
 * 16 lowercase hex chars, well within the slug regex
 * (`^[a-z0-9](?:[a-z0-9-]{0,78}[a-z0-9])?$`) used by the artifact store.
 */

const HEX = '0123456789abcdef'

function hexFromUint32(n: number): string {
  // Unsigned 32-bit integer -> 8 hex chars.
  const u = n >>> 0
  let out = ''
  for (let shift = 28; shift >= 0; shift -= 4) {
    out += HEX[(u >> shift) & 0xf]
  }
  return out
}

/**
 * Compute a 16-hex-char deterministic slug from a message timestamp and
 * a widget's 0-based index within that message. Two FNV-1a passes with
 * the standard 32-bit prime but different starting offset bases give us
 * 64 bits of namespace at zero crypto cost.
 *
 * Note: we use the 32-bit FNV prime (0x01000193) for BOTH passes, not
 * the 64-bit prime (0x100000001b3). Math.imul truncates its operands to
 * 32-bit signed integers before multiplying, so the 64-bit prime would
 * silently get truncated to 0x1b3 (435 decimal) — collapsing the second
 * pass to a weak `multiply-by-435` hash with poor avalanche behavior.
 * The two different offset bases (0x811c9dc5 and 0x62b82175) are enough
 * to guarantee independence of the two passes given the same input.
 *
 * @param messageTs Slack-style timestamp (e.g. "1779995123.456789") or any
 *                  string identifier for the parent message.
 * @param widgetIndex 0-based ordinal of the widget within the message.
 */
export function deriveWidgetSlug(messageTs: string, widgetIndex: number): string {
  const seed = `${messageTs}#${widgetIndex}`
  // Two independent FNV-1a passes — 32-bit prime, different offset bases.
  let h1 = 0x811c9dc5 >>> 0
  let h2 = 0x62b82175 >>> 0
  for (let i = 0; i < seed.length; i++) {
    const c = seed.charCodeAt(i)
    h1 = Math.imul(h1 ^ c, 0x01000193) >>> 0
    h2 = Math.imul(h2 ^ c, 0x01000193) >>> 0
  }
  return hexFromUint32(h1) + hexFromUint32(h2)
}

/**
 * Pick the effective slug for a widget impression — explicit attribute
 * wins; otherwise derived from message location. Returns null only when
 * neither input is available (e.g. a streaming or detached widget that
 * has no parent message context yet).
 */
export function effectiveWidgetSlug(opts: {
  explicitSlug?: string | null
  messageTs?: string | null
  widgetIndex?: number | null
}): string | null {
  if (opts.explicitSlug) return opts.explicitSlug
  if (opts.messageTs && typeof opts.widgetIndex === 'number') {
    return deriveWidgetSlug(opts.messageTs, opts.widgetIndex)
  }
  return null
}

File: src/kiro_crew/widget_slug.py
"""Deterministic widget-artifact slug derivation (Python side of a two-language pair).

Every ``<mcwidget>`` emitted in chat is auto-registered as an artifact keyed by a
slug derived from its location in the conversation — the parent message timestamp
plus the widget's 0-based ordinal within that message. The frontend
(``website/src/lib/widgetSlug.ts``) derives the SAME slug from the SAME inputs so
a ``WidgetFrame`` impression can look up the artifact the backend registered
without the two sides exchanging an id.

**This file and ``widgetSlug.ts`` MUST stay byte-identical in output.** If they
diverge, every auto-registered widget becomes invisible to the frontend (the
probe 404s) and the star button silently creates a duplicate at the frontend's
slug — the exact save-then-refresh duplication the deterministic scheme exists to
prevent. ``test/test_widget_slug.py`` pins known input -> output vectors, and
``website/src/test/widgetSlug.test.ts`` asserts the same vectors on the TS side;
change one without the other and both suites fail.

The hash is two FNV-1a passes with the 32-bit prime (``0x01000193``) and two
different offset bases, emitted as 16 lowercase hex chars. No crypto properties
are needed — this is an opaque unique-enough id, not a security boundary.

Note the 32-bit prime is used for BOTH passes, deliberately: JavaScript's
``Math.imul`` truncates operands to 32-bit before multiplying, so the 64-bit FNV
prime would silently collapse to ``0x1b3`` there. Python has no such limit, so
matching the JS behavior means explicitly masking to 32 bits here and using the
same prime.
"""

from __future__ import annotations

import re

#: FNV-1a 32-bit prime, used for both passes (see module docstring).
_FNV_PRIME = 0x01000193

#: Offset bases for the two independent passes. Two different bases give 64 bits
#: of namespace from one prime.
_FNV_OFFSET_A = 0x811C9DC5
_FNV_OFFSET_B = 0x62B82175

#: 32-bit mask — mirrors the implicit truncation of JS ``Math.imul`` + ``>>> 0``.
_U32 = 0xFFFFFFFF

#: A derived slug is exactly 16 lowercase hex chars (two 32-bit words). Used to
#: recognize (and re-derive) auto-registered widget slugs.
DERIVED_SLUG_RE = re.compile(r"^[0-9a-f]{16}$")


def derive_widget_slug(message_ts: str, widget_index: int) -> str:
    """Return the deterministic 16-hex-char slug for a widget impression.

    ``message_ts`` is the parent message's timestamp (any stable string id);
    ``widget_index`` is the widget's 0-based ordinal within that message. Two
    widgets in one message differ only by the index.

    Must match ``deriveWidgetSlug`` in ``website/src/lib/widgetSlug.ts``.
    """
    seed = f"{message_ts}#{widget_index}"
    h1 = _FNV_OFFSET_A
    h2 = _FNV_OFFSET_B
    for ch in seed:
        code = ord(ch)
        # JS iterates UTF-16 code units via charCodeAt. Python iterates code
        # points, so an astral character (one code point, two surrogates in JS)
        # would hash differently. Split it into surrogates to match JS exactly.
        if code > 0xFFFF:
            offset = code - 0x10000
            for unit in (0xD800 + (offset >> 10), 0xDC00 + (offset & 0x3FF)):
                h1 = ((h1 ^ unit) * _FNV_PRIME) & _U32
                h2 = ((h2 ^ unit) * _FNV_PRIME) & _U32
            continue
        h1 = ((h1 ^ code) * _FNV_PRIME) & _U32
        h2 = ((h2 ^ code) * _FNV_PRIME) & _U32
    return f"{h1:08x}{h2:08x}"
